test/publicresolver_test.js

```diff
@@ -1,85 +1,384 @@
-			function(done) {
-				resolver = web3.eth.contract(JSON.parse(resolverCode.interface)).new(
-				    ens.address,
-				    {
-				    	from: accounts[0],
-				     	data: resolverCode.bytecode,
-				     	gas: 4700000
-				   	}, function(err, contract) {
-				   	    assert.equal(err, null, err);
-				   	    if(contract.address != undefined) {
-				   	    	ens.setSubnodeOwner(0, web3.sha3('eth'), accounts[0], {from: accounts[0]}, done);
-					   	}
-				   });
-			}],
-			done
-		);
-	});
+	beforeEach(function() {
+		this.timeout(4000);
+		return utils.deployENSAsync(accounts[0])
+			.then(_ens => {
+				ens = _ens;
+				return new Promise(function(resolve, reject) {
+					web3.eth.contract(JSON.parse(resolverCode.interface)).new(
+						ens.address,
+						{
+							from: accounts[0],
+							data: resolverCode.bytecode,
+							gas: 4700000
+						},
+						function(err, contract) {
+							if (err) {
+								reject(err);
+							} else if (typeof contract.address !== "undefined") {
+								resolve(contract);
+							} else {
+								// There is to hope that reject or resolve is called
+							}
+						});
 
-	it('permits setting addresses', function(done) {
-		async.series([
-			function(done) { resolver.supportsInterface("0x3b3b57de", function(err, result) {
-					assert.equal(err, null, err);
-					assert.equal(result, true);
-					done();
-				});
-			},
-			function(done) { resolver.setAddr(utils.node, accounts[1], {from: accounts[0]}, function(err, txid) {
-					assert.equal(err, null, err);
-					done();
 				});
-			},
-			function(done) { resolver.addr(utils.node, function(err, address) {
-					assert.equal(err, null, err);
-					assert.equal(address, accounts[1]);
-					done();
+			})
+			.then(contract => {
+				resolver = Promise.promisifyAll(contract);
+				ens.setSubnodeOwnerAsync(0, web3.sha3('eth'), accounts[0], {from: accounts[0]});
+			});
+	});
+
+	describe('constructor', function() {
+
+		it('should not allow a 0 address for ens');
+
+		it.only('uses precise gas', function() {
+			return web3.eth.getTransactionReceiptAsync(resolver.transactionHash)
+				.then(receipt => assert.equal(receipt.gasUsed, 349348));
+		});
+
+	});
+
+	describe('fallback function', function() {
+
+		it('forbids calls to the fallback function with 0 value', function() {
+			return web3.eth.sendTransactionAsync({
+					from: accounts[0],
+					to: resolver.address,
+					gas: 3000000
+				})
+				.then(
+					tx => { throw new Error("expected to be forbidden"); },
+					err => assert.ok(err, err)
+				);
+		});
+
+		it('forbids calls to the fallback function with 1 value', function() {
+			return web3.eth.sendTransactionAsync({
+					from: accounts[0],
+					to: resolver.address,
+					gas: 3000000,
+					value: 1
+				})
+				.then(
+					tx => { throw new Error("expected to be forbidden"); },
+					err => assert.ok(err, err)
+				);
+		});
+
+	});
+
+	describe('has function', function() {
```

Comment:
I have some bad news: `has` is obsolete, and is left in for legacy purposes. Thanks for testing it so thoroughly, though! :)